### ISSUE TITLE
go-pluginserver termianted when upgrade to 2.3.1

### ISSUE THREAD

**Issue by Ackerr** [Kong/kong#6792]
### Summary
 
Upgrade kong2.1 to kong2.3, according to [the document](https://docs.konghq.com/2.3.x/external-plugins/#kong-gateway-plugin-server-configuration), I add some new  pluginserver config, now config like this

```
KONG_PLUGINS=bundled,go-auth
KONG_GO_PLUGINS_DIR=/go-plugins
KONG_PLUGINSERVER_NAMES=go
KONG_PLUGINSERVER_GO_SOCKET=go_pluginserver.sock
KONG_PLUGINSERVER_GO_START_CMD="go-pluginserver -kong-prefix /usr/local/kong/ -plugins-directory /go-plugins"
KONG_PLUGINSERVER_GO_QUERY_CMD="go-pluginserver -dump-all-plugins -plugins-directory /go-plugins"
```

But when kong start,  it output some error log 
```
2021/02/01 06:28:52 [notice] 27#0: *22 [kong] process.lua:249 external pluginserver 'go' terminated: exit 127, context: ngx.timer
2021/02/01 06:28:52 [notice] 27#0: *22 [kong] process.lua:239 Starting go, context: ngx.timer
2021/02/01 06:28:52 [notice] 27#0: signal 17 (SIGCHLD) received from 14003
```

And I rollback kong to 2.2.1, it‘s OK.

#### Dockerfile

**Plugin** 

use `go:1.15-alpine`，`go-pluginserver@v0.6.0` and  `go-pdk@v0.6.0`
```
go build -buildmode plugin -o build/go-auth.so plugins/auth/go-auth.go
go get -d -v github.com/Kong/go-pluginserver
go build -o build/go-pluginserver github.com/Kong/go-pluginserver
```

**Kong**
```
FROM kong:2.3.1-alpine

USER kong

COPY build/go-auth.so /go-plugins/go-auth.so
COPY build/go-pluginserver /usr/local/bin/
```

### Additional Details & Logs

- Kong version (`$ kong version`) 2.3.1
- Kong debug-level startup logs (`$ kong start --vv`)  
```
2021/02/01 06:54:45 [verbose] Kong: 2.3.1
2021/02/01 06:54:45 [warn] the 'client_max_body_size' configuration property is deprecated, use 'nginx_http_client_max_body_size' instead
2021/02/01 06:54:45 [debug] ngx_lua: 10017
2021/02/01 06:54:45 [debug] nginx: 1017008
2021/02/01 06:54:45 [debug] Lua: LuaJIT 2.1.0-beta3
2021/02/01 06:54:45 [verbose] no config file found at /etc/kong/kong.conf
2021/02/01 06:54:45 [verbose] no config file found at /etc/kong.conf
2021/02/01 06:54:45 [verbose] no config file, skip loading
2021/02/01 06:54:45 [debug] reading environment variables
2021/02/01 06:54:45 [debug] KONG_PLUGINSERVER_NAMES ENV found with "go"
2021/02/01 06:54:45 [debug] KONG_ADMIN_LISTEN ENV found with "0.0.0.0:8001"
2021/02/01 06:54:45 [debug] KONG_PROXY_ACCESS_LOG ENV found with "/dev/stdout main"
2021/02/01 06:54:45 [debug] KONG_PLUGINSERVER_GO_START_CMD ENV found with "go-pluginserver -kong-prefix /usr/local/kong/ -plugins-directory /go-plugins"
2021/02/01 06:54:45 [debug] KONG_PROXY_LISTEN ENV found with "0.0.0.0:8000, 0.0.0.0:9000 http2"
2021/02/01 06:54:45 [debug] KONG_STATUS_LISTEN ENV found with "0.0.0.0:9254"
2021/02/01 06:54:45 [debug] KONG_PG_DATABASE ENV found with "kong"
2021/02/01 06:54:45 [debug] KONG_NGINX_HTTP_LOG_FORMAT ENV found with "main
  escape=json '$remote_addr - [$time_local] $http_host req="$request" $status $body_bytes_sent
  "$http_referer" "$http_user_agent" $upstream_status upstream=$upstream_addr rt=$request_time
  uht=$upstream_header_time urt=$upstream_response_time req_body="$request_body"' "
2021/02/01 06:54:45 [debug] KONG_NGINX_PROXY_INCLUDE ENV found with "/kong/proxy.nginx.conf"
2021/02/01 06:54:45 [debug] KONG_PLUGINSERVER_GO_QUERY_CMD ENV found with "go-pluginserver -dump-all-plugins -plugins-directory /go-plugins"
2021/02/01 06:54:45 [debug] KONG_PG_PORT ENV found with "3433"
2021/02/01 06:54:45 [debug] KONG_ERROR_DEFAULT_TYPE ENV found with "application/json"
2021/02/01 06:54:45 [debug] KONG_TRUSTED_IPS ENV found with "0.0.0.0/0"
2021/02/01 06:54:45 [debug] KONG_PG_USER ENV found with "kong"
2021/02/01 06:54:45 [debug] KONG_PLUGINSERVER_GO_SOCKET ENV found with "go_pluginserver.sock"
2021/02/01 06:54:45 [debug] KONG_PG_HOST ENV found with "host"
2021/02/01 06:54:45 [debug] KONG_GO_PLUGINS_DIR ENV found with "/go-plugins"
2021/02/01 06:54:45 [debug] KONG_DATABASE ENV found with "postgres"
2021/02/01 06:54:45 [debug] KONG_ADMIN_ACCESS_LOG ENV found with "/dev/null"
2021/02/01 06:54:45 [debug] KONG_PROXY_ERROR_LOG ENV found with "/dev/stderr"
2021/02/01 06:54:45 [debug] KONG_CLIENT_MAX_BODY_SIZE ENV found with "100m"
2021/02/01 06:54:45 [debug] KONG_ADMIN_ERROR_LOG ENV found with "/dev/null"
2021/02/01 06:54:45 [debug] KONG_REAL_IP_HEADER ENV found with "X-Forwarded-For"
2021/02/01 06:54:45 [debug] KONG_PG_PASSWORD ENV found with "******"
2021/02/01 06:54:45 [debug] KONG_ANONYMOUS_REPORTS ENV found with "off"
2021/02/01 06:54:45 [debug] KONG_PLUGINS ENV found with "bundled,go-auth"
2021/02/01 06:54:45 [debug] admin_access_log = "/dev/null"
2021/02/01 06:54:45 [debug] admin_error_log = "/dev/null"
2021/02/01 06:54:45 [debug] admin_listen = {"0.0.0.0:8001"}
2021/02/01 06:54:45 [debug] admin_ssl_cert = {}
2021/02/01 06:54:45 [debug] admin_ssl_cert_key = {}
2021/02/01 06:54:45 [debug] anonymous_reports = false
2021/02/01 06:54:45 [debug] cassandra_contact_points = {"127.0.0.1"}
2021/02/01 06:54:45 [debug] cassandra_data_centers = {"dc1:2","dc2:3"}
2021/02/01 06:54:45 [debug] cassandra_keyspace = "kong"
2021/02/01 06:54:45 [debug] cassandra_lb_policy = "RequestRoundRobin"
2021/02/01 06:54:45 [debug] cassandra_port = 9042
2021/02/01 06:54:45 [debug] cassandra_read_consistency = "ONE"
2021/02/01 06:54:45 [debug] cassandra_refresh_frequency = 60
2021/02/01 06:54:45 [debug] cassandra_repl_factor = 1
2021/02/01 06:54:45 [debug] cassandra_repl_strategy = "SimpleStrategy"
2021/02/01 06:54:45 [debug] cassandra_schema_consensus_timeout = 10000
2021/02/01 06:54:45 [debug] cassandra_ssl = false
2021/02/01 06:54:45 [debug] cassandra_ssl_verify = false
2021/02/01 06:54:45 [debug] cassandra_timeout = 5000
2021/02/01 06:54:45 [debug] cassandra_username = "kong"
2021/02/01 06:54:45 [debug] cassandra_write_consistency = "ONE"
2021/02/01 06:54:45 [debug] client_body_buffer_size = "8k"
2021/02/01 06:54:45 [debug] client_max_body_size = "100m"
2021/02/01 06:54:45 [debug] client_ssl = false
2021/02/01 06:54:45 [debug] cluster_control_plane = "127.0.0.1:8005"
2021/02/01 06:54:45 [debug] cluster_data_plane_purge_delay = 1209600
2021/02/01 06:54:45 [debug] cluster_listen = {"0.0.0.0:8005"}
2021/02/01 06:54:45 [debug] cluster_mtls = "shared"
2021/02/01 06:54:45 [debug] database = "postgres"
2021/02/01 06:54:45 [debug] db_cache_ttl = 0
2021/02/01 06:54:45 [debug] db_cache_warmup_entities = {"services"}
2021/02/01 06:54:45 [debug] db_resurrect_ttl = 30
2021/02/01 06:54:45 [debug] db_update_frequency = 5
2021/02/01 06:54:45 [debug] db_update_propagation = 0
2021/02/01 06:54:45 [debug] dns_error_ttl = 1
2021/02/01 06:54:45 [debug] dns_hostsfile = "/etc/hosts"
2021/02/01 06:54:45 [debug] dns_no_sync = false
2021/02/01 06:54:45 [debug] dns_not_found_ttl = 30
2021/02/01 06:54:45 [debug] dns_order = {"LAST","SRV","A","CNAME"}
2021/02/01 06:54:45 [debug] dns_resolver = {}
2021/02/01 06:54:45 [debug] dns_stale_ttl = 4
2021/02/01 06:54:45 [debug] error_default_type = "application/json"
2021/02/01 06:54:45 [debug] go_plugins_dir = "/go-plugins"
2021/02/01 06:54:45 [debug] go_pluginserver_exe = "/usr/local/bin/go-pluginserver"
2021/02/01 06:54:45 [debug] headers = {"server_tokens","latency_tokens"}
2021/02/01 06:54:45 [debug] host_ports = {}
2021/02/01 06:54:45 [debug] kic = false
2021/02/01 06:54:45 [debug] log_level = "notice"
2021/02/01 06:54:45 [debug] lua_package_cpath = ""
2021/02/01 06:54:45 [debug] lua_package_path = "./?.lua;./?/init.lua;"
2021/02/01 06:54:45 [debug] lua_socket_pool_size = 30
2021/02/01 06:54:45 [debug] lua_ssl_trusted_certificate = {}
2021/02/01 06:54:45 [debug] lua_ssl_verify_depth = 1
2021/02/01 06:54:45 [debug] mem_cache_size = "128m"
2021/02/01 06:54:45 [debug] nginx_admin_client_body_buffer_size = "10m"
2021/02/01 06:54:45 [debug] nginx_admin_client_max_body_size = "10m"
2021/02/01 06:54:45 [debug] nginx_admin_directives = {{name="client_body_buffer_size",value="10m"},{name="client_max_body_size",value="10m"}}
2021/02/01 06:54:45 [debug] nginx_daemon = "on"
2021/02/01 06:54:45 [debug] nginx_events_directives = {{name="multi_accept",value="on"},{name="worker_connections",value="auto"}}
2021/02/01 06:54:45 [debug] nginx_events_multi_accept = "on"
2021/02/01 06:54:45 [debug] nginx_events_worker_connections = "auto"
2021/02/01 06:54:45 [debug] nginx_http_client_body_buffer_size = "8k"
2021/02/01 06:54:45 [debug] nginx_http_client_max_body_size = "100m"
2021/02/01 06:54:45 [debug] nginx_http_directives = {{name="ssl_dhparam",value="ffdhe2048"},{name="client_max_body_size",value="100m"},{name="ssl_prefer_server_ciphers",value="off"},{name="log_format",value=[[main
  escape=json '$remote_addr - [$time_local] $http_host req="$request" $status $body_bytes_sent
  "$http_referer" "$http_user_agent" $upstream_status upstream=$upstream_addr rt=$request_time
  uht=$upstream_header_time urt=$upstream_response_time req_body="$request_body"' ]]},{name="client_body_buffer_size",value="8k"},{name="ssl_protocols",value="TLSv1.2 TLSv1.3"},{name="ssl_session_timeout",value="1d"},{name="ssl_session_tickets",value="on"}}
2021/02/01 06:54:45 [debug] nginx_http_log_format = [[main
  escape=json '$remote_addr - [$time_local] $http_host req="$request" $status $body_bytes_sent
  "$http_referer" "$http_user_agent" $upstream_status upstream=$upstream_addr rt=$request_time
  uht=$upstream_header_time urt=$upstream_response_time req_body="$request_body"' ]]
2021/02/01 06:54:45 [debug] nginx_http_ssl_dhparam = "ffdhe2048"
2021/02/01 06:54:45 [debug] nginx_http_ssl_prefer_server_ciphers = "off"
2021/02/01 06:54:45 [debug] nginx_http_ssl_protocols = "TLSv1.2 TLSv1.3"
2021/02/01 06:54:45 [debug] nginx_http_ssl_session_tickets = "on"
2021/02/01 06:54:45 [debug] nginx_http_ssl_session_timeout = "1d"
2021/02/01 06:54:45 [debug] nginx_http_status_directives = {}
2021/02/01 06:54:45 [debug] nginx_http_upstream_directives = {}
2021/02/01 06:54:45 [debug] nginx_main_daemon = "on"
2021/02/01 06:54:45 [debug] nginx_main_directives = {{name="daemon",value="on"},{name="worker_rlimit_nofile",value="auto"},{name="worker_processes",value="auto"}}
2021/02/01 06:54:45 [debug] nginx_main_worker_processes = "auto"
2021/02/01 06:54:45 [debug] nginx_main_worker_rlimit_nofile = "auto"
2021/02/01 06:54:45 [debug] nginx_optimizations = true
2021/02/01 06:54:45 [debug] nginx_proxy_directives = {{name="include",value="/kong/proxy.nginx.conf"},{name="real_ip_header",value="X-Forwarded-For"},{name="real_ip_recursive",value="off"}}
2021/02/01 06:54:45 [debug] nginx_proxy_include = "/kong/proxy.nginx.conf"
2021/02/01 06:54:45 [debug] nginx_proxy_real_ip_header = "X-Forwarded-For"
2021/02/01 06:54:45 [debug] nginx_proxy_real_ip_recursive = "off"
2021/02/01 06:54:45 [debug] nginx_sproxy_directives = {}
2021/02/01 06:54:45 [debug] nginx_status_directives = {}
2021/02/01 06:54:45 [debug] nginx_stream_directives = {{name="ssl_dhparam",value="ffdhe2048"},{name="ssl_session_timeout",value="1d"},{name="ssl_protocols",value="TLSv1.2 TLSv1.3"},{name="ssl_prefer_server_ciphers",value="off"},{name="ssl_session_tickets",value="on"}}
2021/02/01 06:54:45 [debug] nginx_stream_ssl_dhparam = "ffdhe2048"
2021/02/01 06:54:45 [debug] nginx_stream_ssl_prefer_server_ciphers = "off"
2021/02/01 06:54:45 [debug] nginx_stream_ssl_protocols = "TLSv1.2 TLSv1.3"
2021/02/01 06:54:45 [debug] nginx_stream_ssl_session_tickets = "on"
2021/02/01 06:54:45 [debug] nginx_stream_ssl_session_timeout = "1d"
2021/02/01 06:54:45 [debug] nginx_supstream_directives = {}
2021/02/01 06:54:45 [debug] nginx_upstream_directives = {}
2021/02/01 06:54:45 [debug] nginx_worker_processes = "auto"
2021/02/01 06:54:45 [debug] pg_database = "kong"
2021/02/01 06:54:45 [debug] pg_host = "host"
2021/02/01 06:54:45 [debug] pg_max_concurrent_queries = 0
2021/02/01 06:54:45 [debug] pg_password = "******"
2021/02/01 06:54:45 [debug] pg_port = 3433
2021/02/01 06:54:45 [debug] pg_ro_ssl = false
2021/02/01 06:54:45 [debug] pg_ro_ssl_verify = false
2021/02/01 06:54:45 [debug] pg_semaphore_timeout = 60000
2021/02/01 06:54:45 [debug] pg_ssl = false
2021/02/01 06:54:45 [debug] pg_ssl_verify = false
2021/02/01 06:54:45 [debug] pg_timeout = 5000
2021/02/01 06:54:45 [debug] pg_user = "kong"
2021/02/01 06:54:45 [debug] plugins = {"bundled","go-auth"}
2021/02/01 06:54:45 [debug] pluginserver_go_query_cmd = "go-pluginserver -dump-all-plugins -plugins-directory /go-plugins"
2021/02/01 06:54:45 [debug] pluginserver_go_socket = "go_pluginserver.sock"
2021/02/01 06:54:45 [debug] pluginserver_go_start_cmd = "go-pluginserver -kong-prefix /usr/local/kong/ -plugins-directory /go-plugins"
2021/02/01 06:54:45 [debug] pluginserver_names = {"go"}
2021/02/01 06:54:45 [debug] port_maps = {}
2021/02/01 06:54:45 [debug] prefix = "/usr/local/kong/"
2021/02/01 06:54:45 [debug] proxy_access_log = "/dev/stdout main"
2021/02/01 06:54:45 [debug] proxy_error_log = "/dev/stderr"
2021/02/01 06:54:45 [debug] proxy_listen = {"0.0.0.0:8000","0.0.0.0:9000 http2"}
2021/02/01 06:54:45 [debug] real_ip_header = "X-Forwarded-For"
2021/02/01 06:54:45 [debug] real_ip_recursive = "off"
2021/02/01 06:54:45 [debug] role = "traditional"
2021/02/01 06:54:45 [debug] ssl_cert = {}
2021/02/01 06:54:45 [debug] ssl_cert_key = {}
2021/02/01 06:54:45 [debug] ssl_cipher_suite = "intermediate"
2021/02/01 06:54:45 [debug] ssl_ciphers = "ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384"
2021/02/01 06:54:45 [debug] ssl_dhparam = "ffdhe2048"
2021/02/01 06:54:45 [debug] ssl_prefer_server_ciphers = "on"
2021/02/01 06:54:45 [debug] ssl_protocols = "TLSv1.1 TLSv1.2 TLSv1.3"
2021/02/01 06:54:45 [debug] ssl_session_tickets = "on"
2021/02/01 06:54:45 [debug] ssl_session_timeout = "1d"
2021/02/01 06:54:45 [debug] status_access_log = "off"
2021/02/01 06:54:45 [debug] status_error_log = "logs/status_error.log"
2021/02/01 06:54:45 [debug] status_listen = {"0.0.0.0:9254"}
2021/02/01 06:54:45 [debug] status_ssl_cert = {}
2021/02/01 06:54:45 [debug] status_ssl_cert_key = {}
2021/02/01 06:54:45 [debug] stream_listen = {"off"}
2021/02/01 06:54:45 [debug] trusted_ips = {"0.0.0.0/0"}
2021/02/01 06:54:45 [debug] untrusted_lua = "sandbox"
2021/02/01 06:54:45 [debug] untrusted_lua_sandbox_environment = {}
2021/02/01 06:54:45 [debug] untrusted_lua_sandbox_requires = {}
2021/02/01 06:54:45 [debug] upstream_keepalive_idle_timeout = 60
2021/02/01 06:54:45 [debug] upstream_keepalive_max_requests = 100
2021/02/01 06:54:45 [debug] upstream_keepalive_pool_size = 60
2021/02/01 06:54:45 [debug] worker_consistency = "strict"
2021/02/01 06:54:45 [debug] worker_state_update_frequency = 5
2021/02/01 06:54:45 [verbose] prefix in use: /usr/local/kong
2021/02/01 06:54:45 [verbose] preparing nginx prefix directory at /usr/local/kong
2021/02/01 06:54:45 [debug] searching for OpenResty 'nginx' executable
2021/02/01 06:54:45 [debug] /usr/local/openresty/nginx/sbin/nginx -v: 'nginx version: openresty/1.17.8.2'
2021/02/01 06:54:45 [debug] found OpenResty 'nginx' executable at /usr/local/openresty/nginx/sbin/nginx
2021/02/01 06:54:45 [debug] testing nginx configuration: KONG_NGINX_CONF_CHECK=true /usr/local/openresty/nginx/sbin/nginx -t -p /usr/local/kong -c nginx.conf
2021/02/01 06:54:45 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2021/02/01 06:54:45 [debug] kill -0 `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2021/02/01 06:54:45 [debug] loading subsystems migrations...
2021/02/01 06:54:45 [verbose] retrieving database schema state...
2021/02/01 06:54:45 [verbose] schema state retrieved
2021/02/01 06:54:45 [debug] searching for OpenResty 'nginx' executable
2021/02/01 06:54:45 [debug] /usr/local/openresty/nginx/sbin/nginx -v: 'nginx version: openresty/1.17.8.2'
2021/02/01 06:54:45 [debug] found OpenResty 'nginx' executable at /usr/local/openresty/nginx/sbin/nginx
2021/02/01 06:54:45 [debug] sending signal to pid at: /usr/local/kong/pids/nginx.pid
2021/02/01 06:54:45 [debug] kill -0 `cat /usr/local/kong/pids/nginx.pid` >/dev/null 2>&1
2021/02/01 06:54:45 [debug] starting nginx: /usr/local/openresty/nginx/sbin/nginx -p /usr/local/kong -c nginx.conf
2021/02/01 06:54:46 [debug] nginx started
2021/02/01 06:54:46 [info] Kong started

```

- Kong error logs (`<KONG_PREFIX>/logs/error.log`): empty
- Kong configuration: default kong config
- Operating system : kong-2.3.1-alpine

**Comment by javierguerragiraldez** [comment]
We've had lots of problems with environment compatibility for dynamically-linked Go programs, especially with musl libraries like on Alpine Linux.

Some users have reported an easier experience with the [embedded server](https://docs.konghq.com/2.3.x/external-plugins/#embedded-server), which gives you a standalone microservice for each plugin.

**Comment by Ackerr** [comment]
Thanks for your reply. I try the embedded server. It seems work well and not error.

**Comment by anton-mesnyankin** [comment]
I have the same issue, the fix is (tested on 2.4 & 2.3):
```
KONG_PLUGINSERVER_GO_START_CMD="/usr/local/bin/go-pluginserver -kong-prefix /usr/local/kong/ -plugins-directory /go-plugins"
KONG_PLUGINSERVER_GO_QUERY_CMD="/usr/local/bin/go-pluginserver -dump-all-plugins -plugins-directory /go-plugins"
```
I hope that helped someone.
